### ISSUE TITLE
`StringLiteral.getLiteralValue` is not thread-safe

### ISSUE THREAD

**Issue by martinlippert** [eclipse-jdt/eclipse.jdt.core#3424]
It looks like `StringLiteral.getLiteralValue` is throwing `IllegalStateExceptions` in some cases when used from multiple threads, most likely due to the internal scanner object being used in a stateful way.

**Comment by jukzi** [comment]
Please paste the Exception Stacktrace and why it is used in multiple threads

**Comment by martinlippert** [comment]
Example stack traces can be found here: https://github.com/spring-projects/sts4/issues/1434 - but this comes from the Spring Tools using JDT Core to parse and analyze source code, so the stack traces have no origins in JDT or Eclipse itself.

We use the AST over there to implement content-assist, do additional validations, and extract index information from a Spring perspective.

**Comment by jukzi** [comment]
That reference only shows a IllegalArgumentException, not a IllegalStateException , also the line number does not match master.
The IllegalArgumentException complains tokenType!=TerminalTokens.TokenNameStringLiteral

I don't see a connection to concurrent use. You should be able to set a breakpoint to find the sourcecode in question that cannot be parsed.

**Comment by martinlippert** [comment]
> That reference only shows a IllegalArgumentException, not a IllegalStateException , also the line number does not match master. The IllegalArgumentException complains tokenType!=TerminalTokens.TokenNameStringLiteral

You are right, my bad, it is an `IllegalArgumentException`, not an `IllegalStateException`. The line numbers are from using JDT Core 3.39.

> I don't see a connection to concurrent use. You should be able to set a breakpoint to find the sourcecode in question that cannot be parsed.

I am not yet convinced about this not being related to concurrent use, but let me try to dig deeper here. Always hard to reproduce those things, but let me see what I can find out with our hints here. Thanks so much for looking into this so quickly and for providing the feedback here.

**Comment by jukzi** [comment]
This issue still has not enough information to process.

**Comment by martinlippert** [comment]
@martinlippert I am on it and will provide more details / information as soon as possible... :-)

**Comment by martinlippert** [comment]
I created a code snippet to reproduce this issue with concurrent calls to `getLiteralValue`:

```
ASTParser parser = ASTParser.newParser(AST.JLS21);
parser.setSource("\"hello\\nworld\"".toCharArray());
parser.setKind(ASTParser.K_EXPRESSION);
ASTNode ast = parser.createAST(null);

final StringLiteral literal = (StringLiteral) ast;

ExecutorService executorService = Executors.newFixedThreadPool(10);
List<CompletableFuture<Void>> futures = new ArrayList<>();
for (int i = 0; i < 100; i++) {
	CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {
		try {
			literal.getLiteralValue();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}, executorService);
	futures.add(future);
}
CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
executorService.shutdown();
```

This should result in a few exceptions being thrown due to issues within `getLiteralValue` due to concurrent access to it.

Looking at the implementation of `getLiteralValue` it seems that the `scanner` object is the culprit. It is used to scan the escaped value in a stateful way, but calls to `getLiteralValue` always use the same scanner object.

**Comment by jukzi** [comment]
was already reported in https://bugs.eclipse.org/bugs/show_bug.cgi?id=508586

**Comment by jukzi** [comment]
as per https://bugs.eclipse.org/bugs/show_bug.cgi?id=508586#c14 the scanner should be cloned. @martinlippert can you contribute a fix?

**Comment by martinlippert** [comment]
@jukzi Here is the PR with the fix: https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3521

I included your commit for the unit test, simplified the test case a bit, and applied both (the test case and the fix) to CharacterValue.charValue as well (same issue).
